Show the FlashAir IP address

odaki opened the pull request:
Displays the FlashAir IP address in the "Support" menu if the card is available.
If no IP address has been assigned yet, "0.0.0.0" will be displayed.

PR #2088 works fine with my FlashAir W-03 32GB.
However, even if a FlashAir IP address is assigned, the IP address is not displayed.
So, instead of the old commented out code, I created a small patch that shows the IP address.

The patch tested with a FlashAir W-03 32GB.

Please merge if you like.

odaki committed:
Show the FlashAir IP address

Displays the FlashAir IP address in the "Support" menu if the card is available.
If no IP address has been assigned yet, "0.0.0.0" will be displayed.

leptun commented:
I’ll look into this PR tomorrow and test it on my 32GB flashair W-04 card.

leptun commented:
Unfortunately, my flashair card's IP address in not detected and is always 0.0.0.0. Looks like the W-04 is not supported

odaki commented:
Thank you for testing.

I also purchased W-04 for testing. When the card arrives, I will test it additionally.
Thank you for your cooperation.

odaki commented:
I'm investigating information about the differences between the W-03 and W-04 specifications, and I found that the commands to read the internal memory seem to be different.
W-03: CMD48
W-04: CMD17 (firmware 4.0.0.01+)
Changing the code to use a different command for each card,
or trying both CMD17 and CMD48, may work with both W-03 and W-04 types.
I will try it later when W-04 arrives.

https://flashair-developers.github.io/website/docs/api/isdio.html
> ファームウェアバージョン 4.00.01+でのSPIモードでのiSDIOは、CMD48, CMD49ではなく、CMD17, CMD24をお使いください。

It meams 
"Please use CMD17, CMD24 instead of CMD48, CMD49 for iSDIO in SPI mode with firmware version 4.00.01+."

Example:
https://flashair-developers.github.io/website/docs/tutorials/arduino/2

>      // send command and argument.
>     if (cardCommand(CMD48, arg) && cardCommand(CMD17, arg)) {
>       error(SD_CARD_ERROR_CMD48);
>       goto fail;
>     }
>

(First, use CMD48, then if it failed use CMD17 instead of.)

leptun commented:
I'll check it on Monday when I have access to the card again. Can you make the required changes and push them to this PR?

It is quite weird that they chose to use CMD17 since that is the regular read block function (readBlock) that doesn't normally provide the count argument.

odaki committed:
Support W-04 firmware 4.0.0.01+

W-04 needs to use  CMD17 instead of CMD48 to retrieve iSDIO register memories.
To supporting both W-03 and W-04 card, I changed the readExt() code to use CMD48 first and use CMD17 if it returned some error.
This method is from the FlashAir developer sample code.
https://flashair-developers.github.io/website/docs/tutorials/arduino/2

odaki commented:
@leptun 
Thank you for your kind offer.
I committed the fix code.
This code worked fine on W-03 32GB.
Please test-evaluate my code if possible.

odaki committed:
M46 "Show the assigned IP address" activated

According to the reprap wiki, M46 has been assigned to display its assigned IP address, but has been disabled.
Now that Toshiba FlashAir IP address processing is working, I activate the M46 code so that I can read the IP address information via serial.

odaki commented:
According to the reprap wiki, M46 has been assigned to display its assigned IP address, but has been disabled.
https://reprap.org/wiki/G-code#M46:_Show_the_assigned_IP_address

Now that Toshiba FlashAir IP address processing is working, I activate the M46 code so that I can read the IP address information via serial.

SD card type:
 Normal --> returns "n/a" always.
 FlashAir --> returns only IP address string (without header message to make parse easy).
                       returns "?Toshiba FlashAir GetIP failed" if some error occured.

This patch is tested with a FlashAir W-03 32G.

odaki commented:
I received W-04 32GB (firmware 4.00.04).
I tested the patch with the new W-04 32GB I received.
This patch works well with W-04 (and W-03).

I recorded the test results as a movie.
https://youtu.be/Qe0KVpxRumA

Please review this patch and test results,
and if the PR is acceptable, please merge it.

Thank you.

leptun commented:
I am not the one to merge PRs, but I’ll approve it if it works for me

leptun reviewed (commented):
Just tested it on my W-04 card. It works as expected.
Please update the PR to the latest MK3 branch as there have been some changes to the documentation and now there are merge conflicts. Be careful with the comments ("/*") as they are easy to confuse.
Also, I'd change how the support menu is rendered as the current method is not that efficient and flickers a bit when scrolling the menu.
Good job for solving the Flashair problem!
![image](https://user-images.githubusercontent.com/17808203/72524866-7f806b00-386b-11ea-84bb-df6b10a959cc.png)

odaki committed:
Merge remote-tracking branch 'upstream/MK3' into flashair_display_ip

leptun commented:
It is not ok. There is an extra `/*` as I told you earlier

odaki committed:
Remove '/*' from M46 source

I forgot to remove the unintended comment start mark when merge.
It has been deleted.

odaki commented:
@leptun
I am glad that it works in your environment.
Thank you very much for helping!

odaki committed:
Merge remote-tracking branch 'upstream/MK3' into flashair_display_ip

odaki committed:
Update reference URL

There was an official website of Toshiba called "FlashAir Developers '', and
there was information necessary for FlashAir development, but since it closed in September 2019.
Instead, I will point an archive site for developers.

odaki committed:
Merge branch 'MK3_3.9.0' into flashair_display_ip

odaki committed:
Merge branch 'MK3_3.9.0' into flashair_display_ip

odaki committed:
Merge remote-tracking branch 'upstream/MK3' into flashair_display_ip

odaki committed:
Merge remote-tracking branch 'upstream/MK3' into flashair_display_ip

odaki committed:
Merge branch 'MK3_3.9.1' into flashair_display_ip

odaki committed:
Merge branch 'MK3' into flashair_display_ip

odaki committed:
Merge branch 'MK3' into flashair_display_ip

odaki committed:
Merge remote-tracking branch 'upstream/MK3' into flashair_display_ip

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -5977,28 +5977,30 @@ if(eSoundMode!=e_SOUND_MODE_SILENT)
+            bool hasIP = card.ToshibaFlashAir_GetIP(ip);
+            if (hasIP) {
+                // SERIAL_PROTOCOLPGM("Toshiba FlashAir current IP: ");
+                SERIAL_PROTOCOL(int(ip[0]));
```
these logging messages can be shortened (in terms of generated code)

odaki replied:
As you pointed out, it should be a simple output.
Thank you for your review.

DRracer reviewed (approved):
approving in order to work on top of this PR for FW 3.10
